core/src/test/java/org/elasticsearch/discovery/DiscoveryModuleTests.java

```diff
@@ -43,49 +58,84 @@
         Settings settings = Settings.EMPTY;
-        DiscoveryModule module = new DiscoveryModule(settings, null, null, Collections.emptyList());
-        assertBinding(module, Discovery.class, ZenDiscovery.class);
+        DiscoveryModule module = newModule(settings, hostsProvider -> null, Collections.emptyList());
+        assertTrue(module.getDiscovery() instanceof ZenDiscovery);
     }
 
     public void testRegisterDiscovery() {
         Settings settings = Settings.builder().put(DiscoveryModule.DISCOVERY_TYPE_SETTING.getKey(), "custom").build();
-        DummyDiscoPlugin plugin = () -> Collections.singletonMap("custom", () -> Collections::emptyList);
-        DiscoveryModule module = new DiscoveryModule(settings, null, null, Collections.singletonList(plugin));
-        module.addDiscoveryType("custom", NoopDiscovery.class);
-        assertBinding(module, Discovery.class, NoopDiscovery.class);
+        DummyDiscoveryPlugin plugin = () -> Collections.singletonMap("custom", NoopDiscovery::new);
+        DiscoveryModule module = newModule(settings, hostsProvider -> null, Collections.singletonList(plugin));
+        assertTrue(module.getDiscovery() instanceof NoopDiscovery);
     }
 
-    public void testHostsProvider() {
-        Settings settings = Settings.builder().put(DiscoveryModule.DISCOVERY_HOSTS_PROVIDER_SETTING.getKey(), "custom").build();
-        final UnicastHostsProvider provider = Collections::emptyList;
-        DummyDiscoPlugin plugin = () -> Collections.singletonMap("custom", () -> provider);
-        DiscoveryModule module = new DiscoveryModule(settings, null, null, Collections.singletonList(plugin));
-        assertInstanceBinding(module, UnicastHostsProvider.class, instance -> instance == provider);
+    public void testUnknownDiscovery() {
+        Settings settings = Settings.builder().put(DiscoveryModule.DISCOVERY_TYPE_SETTING.getKey(), "dne").build();
+        IllegalArgumentException e = expectThrows(IllegalArgumentException.class, () ->
+            newModule(settings, hostsProvider -> null, Collections.emptyList()));
+        assertEquals("Unknown discovery type [dne]", e.getMessage());
     }
 
-    public void testHostsProviderBwc() {
-        Settings settings = Settings.builder().put(DiscoveryModule.DISCOVERY_TYPE_SETTING.getKey(), "custom").build();
+    public void testDuplicateDiscovery() {
+        DummyDiscoveryPlugin plugin1 = () -> Collections.singletonMap("dup", () -> null);
+        DummyDiscoveryPlugin plugin2 = () -> Collections.singletonMap("dup", () -> null);
+        IllegalArgumentException e = expectThrows(IllegalArgumentException.class, () ->
+            newModule(Settings.EMPTY, hostsProvider -> null, Arrays.asList(plugin1, plugin2)));
+        assertEquals("Cannot register discovery type [dup] twice", e.getMessage());
+    }
+
+    public void testHostsProvider() {
+        Settings settings = Settings.builder().put(DiscoveryModule.DISCOVERY_HOSTS_PROVIDER_SETTING.getKey(), "custom").build();
```

Comment:
The default hosts provider (null) is tested by the defaults test (indirectly, since the only thing we get out of the module is Discovery).